src/auth/Crypto.cc

```diff
@@ -348,24 +361,797 @@ int CryptoAES::validate_secret(const bufferptr& secret)
+    bool result = false;
+    SECItem keyItem;
+    CK_AES_CTR_PARAMS ctr_params = {0};
+    SECItem ivItem;
+
+    slot = PK11_GetBestSlot(CKM_AES_CTR, NULL);
+    if (slot) {
+      keyItem.type = siBuffer;
+      keyItem.data = const_cast<uint8_t*>(reinterpret_cast<const uint8_t*>(key.c_str()));
+      keyItem.len = CryptoAES_256_CTR::KEY_SIZE;
+
+      symkey = PK11_ImportSymKey(slot, CKM_AES_CTR, PK11_OriginUnwrap, CKA_UNWRAP, &keyItem, NULL);
+      if (symkey) {
+        static_assert(sizeof(ctr_params.cb) >= CryptoAES_256_CTR::IV_SIZE, "Must fit counter");
+        ctr_params.ulCounterBits = 128;
+        memcpy(&ctr_params.cb, iv.c_str(), CryptoAES_256_CTR::IV_SIZE);
+        ivItem.type = siBuffer;
+        ivItem.data = (unsigned char*)&ctr_params;
+        ivItem.len = sizeof(ctr_params);
+
+        param = PK11_ParamFromIV(CKM_AES_CTR, &ivItem);
+        if (param) {
+          result = true;
+        }
+      }
+    }
+    if (!result) {
+      cleanup();
+      std::ostringstream ss;
+      ss << "Failed to perform AES-CBC encryption: " << PR_GetError();
+      error = ss.str();
+    }
+    return result;
+  }
+
+private:
+  PK11SlotInfo *slot = nullptr;
+  PK11SymKey *symkey = nullptr;
+  SECItem *param = nullptr;
+
+  void cleanup() {
+    if (param) {
+      SECITEM_FreeItem(param, PR_TRUE);
+      param = nullptr;
+    }
+    if (symkey) {
+      PK11_FreeSymKey(symkey);
+      symkey = nullptr;
+    }
+    if (slot) {
+      PK11_FreeSlot(slot);
+      slot = nullptr;
+    }
+  }
+
+  bool ctr_transform(
+      const bufferlist& input,
+      bufferlist& output,
+      std::string *error) const {
+    bool result = false;
+    SECStatus ret;
+    PK11Context *ectx;
+    int written;
+    unsigned int written2;
+    size_t size = input.length();
+    bufferlist _input(input);
+
+    if ((input.length() % CryptoAES_256_CTR::BLOCK_SIZE) != 0) {
+      if (error) {
+        *error = "Input size is not multiple of 16";
+      }
+      return false;
+    }
+
+    ectx = PK11_CreateContextBySymKey(CKM_AES_CTR, CKA_ENCRYPT, symkey, param);
+    if (ectx) {
+      buffer::ptr buf(
+          (size + CryptoAES_256_CTR::KEY_SIZE - 1)
+          / CryptoAES_256_CTR::KEY_SIZE * CryptoAES_256_CTR::KEY_SIZE);
+      ret = PK11_CipherOp(ectx,
+                          (unsigned char*)buf.c_str(), &written, buf.length(),
+                          (unsigned char*)_input.c_str(), size);
+      if (ret == SECSuccess) {
+        ret = PK11_DigestFinal(ectx,
+                               (unsigned char*)buf.c_str() + written, &written2,
+                               buf.length() - written);
+        if (ret == SECSuccess) {
+          buf.set_length(written + written2);
+          output.append(buf);
+          result = true;
+        }
+      }
+      PK11_DestroyContext(ectx, PR_TRUE);
+    }
+    if (result == false) {
+      if (error) {
+        *error = "Failed to perform AES-CTR encryption: " + PR_GetError();
```

Comment:
`PR_GetError()` returns an integer type, so this isn't actually a string concatenation - it's returning a const char pointer plus an offset. you'll need to use the `stringstream` approach as above